### PR TITLE
Billiard walk

### PR DESCRIPTION
**Pull request by TolisChal**
Add the billiard walk for uniform sampling and volume computation:

i) The walk is available for all the representations and provides a speed up for volume computation. 
ii) Its mixing time is better than the existing random walks' mixing time and hence, it is recommended  as the default random walk for uniform sampling.

### CODE REVIEW

File: include/samplers/samplers.h [hunk 1]
```diff
@@ -425,4 +439,51 @@ void hit_and_run_coord_update(Point &p,
     p.set_coord(rand_coord, p[rand_coord] + bpair.first + kapa * (bpair.second - bpair.first));
 }
 
+
+template <class ConvexBody, class Point, class Parameters, typename NT>
+void billiard_walk(ConvexBody &P, Point &p, NT diameter, std::vector<NT> &Ar, std::vector<NT> &Av, NT &lambda_prev,
```
Comment (vissarion) [inline include/samplers/samplers.h]: whatever is not changed by the function should be `const`, e.g. `ConvexBody P` is not changed right?

File: include/convex_bodies/vpolytope.h [hunk 2]
```diff
@@ -365,15 +365,19 @@ class VPolytope{
...
                                                const std::vector<NT> &Av) {
-        return std::pair<NT, int> (intersect_line_Vpoly(V, r, v, conv_comb, row, colno, false, false), 1);
+        return line_positive_intersect(r, v);
     }
 
 
     std::pair<NT, int> line_positive_intersect(const Point &r, const Point &v, const std::vector<NT> &Ar,
                                                const std::vector<NT> &Av, const NT &lambda_prev) {
-        return line_positive_intersect(r, v, Ar, Av);
+        return line_positive_intersect(r, v);//, Ar, Av);
```
Comment (vissarion) [inline include/convex_bodies/vpolytope.h]: if `Ar` and `Av` are not used then remove them from parameters

File: include/samplers/samplers.h [hunk 3]
```diff
@@ -425,4 +439,51 @@ void hit_and_run_coord_update(Point &p,
...
+
+    if (first) {
+
+        std::pair<NT, int> pbpair = P.line_positive_intersect(p, v, Ar, Av);
+        if (T <= pbpair.first) {
+            p = (T * v) + p;
+            lambda_prev = T;
+            return;
+        }
+        lambda_prev = 0.995 * pbpair.first;
```
Comment (vissarion) [inline include/samplers/samplers.h]: `0.995` is this for robustness? maybe define it as  `const` variable

File: R-proj/src/sample_points.cpp [hunk 4]
```diff
@@ -236,6 +246,7 @@ Rcpp::NumericMatrix sample_points(Rcpp::Nullable<Rcpp::Reference> P = R_NilValue
                     InnerBall = HP.ComputeInnerBall();
                     if (!set_mean_point) MeanPoint = InnerBall.first;
                 }
+                if (diam < 0.0) diam = 2.0 * std::sqrt(NT(dim)) * InnerBall.second;
```
Comment (vissarion) [inline R-proj/src/sample_points.cpp]: you set this `std::sqrt(NT(dim)) * InnerBall.second` in various places in the code. It is better if you use everywhere `comp_diam()`  So this is set only in one place in the code. That is, if we want to change it in the future we should only change it there. This applies to all types of convex bodies.

File: include/samplers/samplers.h (hunk 3 above)
Comment (TolisChal) [reply include/samplers/samplers.h]: done

File: R-proj/src/sample_points.cpp (hunk 4 above)
Comment (TolisChal) [reply R-proj/src/sample_points.cpp]: done

**Pull request merged**